### DISCUSSION TITLE
CML Coupons - Mainnet

### DISCUSSION THREAD

**Discussion by geopor** [tearust/teaproject#99] [Ideas]
# Real CML Vouchers Available Starting in Epoch 7

The team is looking into creating vouchers for mainnet CML, which will be released as either a BEP-20 token on the Binance Smart Chain or as a ERC-20 token on the Polkadot network. When the TEA Project mainnet launches, these tokens can be exchanged for real CML at the following rate:

- 1 token = 1 C CML (mainnet)
- 2 token = 1 B CML (mainnet)

The mining contest participants can exchange some of their TEA they win in the contest for these token vouchers. Every week, 2 of these tokens will be released with a valuation of $500 each. 

Starting in epoch 7, the TEA Project will move to a continuous contest model where contestants can do one of three things at the end of each week: 

1. Choose to purchase one of these vouchers once their cumulative contest earnings reach a TEA value exceeding $500.
2. Exchange some or all of the TEA in their wallet for USDT at the exchange rate explained below.
3. Keep the TEA in their wallet. Because the contest is now continuous, the TEA amount in contestants' wallets will continue to accrue in their wallet until they either cash out for USDT or exchange for a CML. 

# Exchanging TEA for CML Vouchers
We will have a set trading day every week where contestants can either cash out their TEA for USDT or exchange their TEA for CML vouchers. The exchange will be done OTC (over the counter) style where we'll manually exchange these tokens with contestants at the exchange rate of 1 (C CML) token / 500 USDT. Except we don't take USDT, we take 500 USDT worth of TEA tokens valued at the prevailing exchange rate that fluctuates as the contest progresses.

# Figuring the TEA->CML Voucher and the TEA->USDT Exchange Rates
To figure the **TEA->USDT** exchange rate, we first find the **TEA->CML Voucher** exchange rate and work back from there.
- 2 tokens are released every week, each valued at $500 each. That means there are $1000 worth of rewards added every week.
- The following formula determines the exchange rate for TEA: (USDT value of the tokens in existence) / (total amount of TEA in existence).

For example, if after the first week there is a total of 10000T and there are 2 tokens valued at a total of $1000, then $1000/10000T = TEA is valued at $.10 for exchange purposes. If a contestant has 1000T, that would be valued at $100 which would be valued at .2 of a token (the tokens are worth $500 each). If this contestant wants to cash out their 1000T:

- They send 1000T to our sudo account.
- The sudo account burns the 1000T as well as the equivalent amount of the voucher token (.2).
- Sudo transfers the USDT equivalent of the burned token (.2 x $500 = $100) to the contestant.

The exchange rate is kept in balance as the contest progresses as both TEA and the CML voucher tokens are burned at the same time.

**Comment by kevingzhang** [comment]
If we can reach our development goal, since epoch7, we will not need to restart our chain every 2 weeks. That means epoch7 could be our latest epoch. The further epoch will just hot update over the existing one. That makes the idea come to reality

**Reply by Charliebrickman75** [reply]
I think liquid staking when feasible would be a great asset addition to your organization as well as themed Nft drops for new users

**Comment by kevingzhang** [comment]
Another idea is to create a DOT NFT. Every NFT has only a CML number in it. Since the CML content (lifespan, performance etc) are generated in the genesis block when TEA main net starts. It is fair to every one. Eg. if you own a CML 200 NFT. You will know you will get a type B CML, but you do not know the life span performance etc yet. It is pure random and fair to every one.

In this case, we can easily give every contestant an NFT in DOT. No need for bridge. What do you think?

**Reply by lxm218** [reply]
I don't think create DOT NFT is a good idea for this purpose. It's just an interim solution. Voucher makes more sense. Try not to make this too complicated.

**Comment by kevingzhang** [comment]
the easiest solution would be create a page on our website. Has a table on it. Everyone owns which CML id. how about that?

**Comment by geopor** [comment]
- We shouldn't be selling CML vouchers using a Google form. And we shouldn't be manually exchanging vouchers for tokens as this leads to weird situations (they have to wait for team to wake up, or some mix up like what's already happened. We shouldn't be inspecting receipts on the blockchain to see if people paid, that's a job for smart contracts). Having to make decisions about whether to take BUSD is just a symptom that we're too hands on. If someone wants to trade BNB for whatever it is we're selling, why should we have any control over that? If liquidity for our token is in BUSDT, then let the pancake swap router take care of mapping their source funds to the target destination. This is crypto and we shouldn't be manually doing anything. It should be autonomous - people should be able to get in and get out of our mining game 24/7.
- Propose that to foster a 24/7 mining game, we only need to have one token that we fund, let's just call it for now Contest-TEA. This is the **TEA** that is currently in the browser wallet. We've already said that we'd fund rewards at $1K per week, we can simply put that amount every week in the Contest-TEA AMM. (it could be on BSC KSM or DOT, there are +/- for each platform). We don't have to worry about the exchange rate, that takes care of itself, that's what AMM is for. Contestants have to purchase Contest-TEA in the AMM in order to buy contest B & C CML or TApp tokens.
- All B & C contest CML are purchased in the existing Marketplace in our TEA browser wallet. It would be a simple API call to get the prevailing exchange rate to set the price of C CML (currently we're asking $7.50) or B CML ($20) in terms of Contest-TEA. If this is too complicated then it's still an easy task for sudo to do once or twice a day.
- We can still sell real mainnet B & C CML vouchers at the proposed price: $500 in equivalent Contest-TEA for mainnet C CML voucher, double that price for B CML voucher. This is something that could be done manually as it would happen infrequently - for example,  a contestant sends us $500 worth of Contest-TEA and we send them a C CML voucher while burning the TEA they send us, same mechanics as mainnet as far as burning TEA.

This system allows for the system to be autonomous 24/7 and allows us to see the dynamics of the TEA token economy unfold, and we have a couple factors that boost the price of Contest-TEA:
- Contestants exchanging Contest-TEA for mainnet vouchers (Contest-TEA is burnt).
- New contestants joining / existing contestants exchanging for more Contest-TEA especially to buy B & C contest CML from the marketplace.

TLDR:
- $1k/week goes into a Contest-TEA automated market maker smart contract.
- Contestants exchange their crypto tokens for Contest-TEA. They then use Contest-TEA to buy B and C contest CML at whatever the prevailing Contest-TEA exchange rate is for our desired price (currently C = $7.50, B CML = $20).
- We can provide token vouchers for anyone who wants to buy mainnet CML vouchers. They would pay the Contest-TEA equivalent of $500 for C, $1000 for B. Team burns the Contest-TEA just like in mainnet.

**Reply by geopor** [reply]
Less manual intervention and more automated, that is very important. It will open us up to more contestants who will finally see us as an actual crypto project and not some contest with a sign up form. This will also help onboard new YouTubers if they know they can sell their TApp tokens to Contest-TEA and cash out 24/7 on the blockchain.

**Reply by lxm218** [reply]
I think it's a good idea to have contestants purchase contest-TEA on swap(e.g. pancake swap) and they are able to purchase CML whenever they want which make the whole manual process easier. My question is if we don't have enough contest TEA put into AMA, the swap won't work properly, is that right?

**Reply by geopor** [reply]
@lxm218 yes that's correct. Let's say that we wanted to allow epoch 6 contestants to cash out via AMM, then we would just take the total TEA generated during epoch 6 (since it started at 0) and put that into the AMM at the end along with $1000 USDT. We could then use this AMM for epoch 7 instead of using a signup form. Not everyone will cash out at the end so some funds will be still in the AMM.

**Reply by lxm218** [reply]
The problem is that we don't have bridge to put contest TEA on BSC or other chain...polkadot doesn't have stable coin

**Reply by Charliebrickman75** [reply]
Automated by smart contract would be much more efficient

**Comment by kevingzhang** [comment]
> 

This is the reason why we introduce COFFEE at the first place. But we cannot use this solution simply because there is no bridge between TEA and other popular chain, (ETH USDT for example).

If we have the bridge, we can start the main net. All those cumbersome should be gone. 
If we do not have the bridge (still waiting for ChainBridge upgrade), all those solutions have to rely on manual process. Of course we should not use manual process but this is our reality without bridge.

Do you have a solution that we can use without a bridge? @geopor

**Reply by Charliebrickman75** [reply]
Kevin, I would love to be a part of your business development team, I think a partnership with a meme token would be very beneficial in bringing on new users. I am a big supporter of Shiba Inu, maybe a mutually themed NFT for the first 1,000 new onboard borrowers?

**Comment by kevingzhang** [comment]
To explain more specificly. Since we do not have a bridge, there is no way for others to buy TEA / COFFEE or any assets in TEA blockchain automatically. The manual process cannot be avoided. The bridge is supposed to do so.

**Comment by geopor** [comment]
@lxm218  Acala has a stablecoin on Polkadot, aUSD and also on Kusama (kUSD).

@kevingzhang 
Polkadot doesn't have smart contracts which is needed to mint new tokens. But we can launch a token on a smart contract parachain like Moonbeam. But since their Polkadot parachain slot is still being approved we can use their Moonriver Kusama parachain.

- Mint the **Contest-TEA** token on Moonriver. This is simple to do.
- There are two cross-chain DEXs on Moonriver, Solarbeam.io and Moonlit.finance. They both integrate the [Anyswap Bridge](https://anyswap.exchange/#/bridge) which can bridge many tokens among many chains but of particular interest to us: USDT (ETH or BSC) <-> USDT (Moonriver). They would just need some Moonriver token (MOVR) for gas, and their bridges have a feature to include a bit of MOVR in the transaction during the bridge transaction.
- Contestants bridge over to Moonriver to mine in the TEA Project. The Contest-TEA is in their wallet, we just read their wallets to calculate their place on the leaderboard, except this Contest-TEA is exchangeable by them at any time. 

I'm not sure what special requirements Solarbeam and Moonlit have for projects wanting to use them for liquidity. One thing I like about ETH and BSC is that anyone can create and list a token on Uniswap / Pancakeswap, it's very open. Polkadot ecosystem seems more gated. I have asked that question to both projects as well as someone who specializes minting tokens on Moonriver, will wait for a response.

**Reply by kevingzhang** [reply]
You can use moonbean or whatever ink contract to create assets on DOT or KSM, but it is NOT in TEA blockchain. There is no bridge so far. The problem is still exists. There is no bridge. I totally understand your expectation. I have the same expectation. but there is NO BRIDGE. We cannot directly use any assets (token) in other chain. no matter it is ETH, BSC or DOT KSM. This is the real problem.

**Reply by Charliebrickman75** [reply]
What about building on a meme token platform like Shibarium to build a mutually beneficial relationship or Polygon

**Reply by kevingzhang** [reply]
> What about building on a meme token platform like Shibarium to build a mutually beneficial relationship or Polygon

I think you are talking about an application (we call them TApps). Yes, you can do whatever you want. TEA supports more types of applications then traditional blockchains can do.

**Reply by Charliebrickman75** [reply]
I would love to join your project

**Comment by geopor** [comment]
What's blocking us is that Chainbridge hasn't updated their Docker file's substrate version, correct? But another similar project has integrated Chainbridge, I saw them in the discord asking questions. They even notated what they did to push it through:

 1)  Fork and upgrade go-substrate-rpc-client in ChainBridge to v3:

go get -u github.com/centrifuge/go-substrate-rpc-client/v3

2) Same steps as 1), fork and upgrade go-substrate-rpc-client in chanbridge-utils and chainbridge-substrate-events.

3) Replace dependences path with our fork repo, like change github.com/ChainSafe/chainbridge-utils to same chainbridge-utils folder in our GitHub repo.

4)  All references to: import "github.com/centrifuge/go-substrate-rpc-client" needs to be replaced with "github.com/centrifuge/go-substrate-rpc-client/v3" within the source code.

5) make build

If our dev team needs someone to relay issues we're having with the Chainbridge team to ask their advice I can help with that.

**Reply by kevingzhang** [reply]
We are using substrate V 14 which supports online update. Chainbridge doesn't upgrade to V 14. so we either drop the online update feature (almost impossible) or wait till Chainbridge upgrade(they will , but do not know when)

**Comment by Charliebrickman75** [comment]
I would love to join the team to assist with business development. Is anyone working on reaching out to an NFT creator.

**Reply by kevingzhang** [reply]
can you reach me kevin.zhang.canada@gmail.com and introduce yourself?
